Compatibility with twig/twig 3 (Case 99556)

MalteWunsch opened the pull request:
- BC break in other areas, as described in changelog
- miscellaneous modernisations

MalteWunsch committed:
Erlaube twig/twig ^3 (Case 99556)

twig/twig wird bereits genamespaced verwendet.

MalteWunsch committed:
Ersetze NavigationTheme-Funktionalität (Case 123872)

Jetzt kann im Projekt direkt das WebfactoryNavigationBundle:Navigation:navigation.html.twig extended und blockweise überschrieben werden.

Beispiel:

{% extends "WebfactoryNavigationBundle:Navigation:navigation.html.twig" %}

{% block navigation_list %}
    <nav class="projektspezifische-wrapperklasse">
        {{ parent() }}
    </nav>
{% endblock %}

MalteWunsch committed:
Verschiebe Historie aus Readme in Changelog (Case 123872)

MalteWunsch committed:
Rename file (Case 123872)

MalteWunsch committed:
Liste die wichtigsten in Blöcken verfügbaren Variablen (Case 123872)

Platziere sie überhalb des jeweiligen Blocks statt darin, weil Kommentare im Block die Whitespace-Control stören / Kommentare mit "-" als Whitespace-Control nicht mehr von der IDE zur Autovervollständigung berücksichtigt werden.

MalteWunsch committed:
Twig-Funktionen statt Controller als Einstiegspunkte (Case 123872)

MalteWunsch committed:
Teile Bedingung auf, reduziere Scope der visibleNodes (Case 123872)

MalteWunsch committed:
Benenne Variable im Einstiegsblock navigation_if_visible wieder root statt node (Case 123872)

MalteWunsch committed:
Add usages section to README (Case 123872)

MalteWunsch committed:
Add short introduction (Case 123872)

MalteWunsch committed:
Add section on modifying the navigation tree (Case 123872)

MalteWunsch committed:
Wende php-cs-fixer an (Case 123872)

MalteWunsch committed:
Fordere direkte Abhängigkeiten (Case 123872)

+ Autoloading nur für eigene Klassen

MalteWunsch committed:
php-cs-fixer as CI workflow (Case 123872)

MalteWunsch committed:
Reduce .gitignore (Case 123872)

+ Ignore php-cs-fixer cache file

mpdude commented on CHANGELOG.md:
````diff
@@ -1,6 +1,81 @@
+    ) }}
+    ```
+    
+    neu: `{{ navigation_ancestry(1) }}` (für weitere Parameter siehe \Webfactory\Bundle\NavigationBundle\Twig\NavigationExtension::renderAncestry)
````
Ich frage mich, ob man das hier sprechender gestalten könnte?

```twig
{{ navigation_ancestry({ startLevel: 1, expandedLevels: 3 }) }}
```

MalteWunsch replied:
Funktioniert out of the box mit named parameters 😃 (ohne die Array/Objekt-Syntax `{}`) :

```
{{ navigation_ancestry(startLevel = 1, template = 'AppBundle:Navigation:secondaryNav.html.twig') }}
```

README und CHANGELOG sind jetzt angepasst, um diese Schreibweise zu empfehlen.

mpdude commented on README.md:
```diff
@@ -1,65 +1,169 @@
+            return;
+        }
+
+        $this->addTreeCacheExpiryRule($dispatcher);
```
"ExpiryRule" verspricht mehr, als dahinter steckt... Symfony's `ConfigCache` nennt die Dinger einfach `Resource`.

MalteWunsch replied:
Ich möchte hier die Intention deutlich machen, da scheint mir `Resource` zu vage. `addResourceForTreeCacheExpiry`? Oder habe ich da schon immer zu viel reininterpretiert - besserer Name?

mpdude replied:
In Symfony heißen diese Methoden überall (Container, MessageCatalogue, Router...) einfach immer `addResource()`. Natürlich muss man wissen, was die `Resource` ist... aber ich weiß nicht, ob da der Methodenname ein hilfreicher clue ist?

MalteWunsch replied:
Die `addResource()`-Methode wird auf dem `Webfactory\Bundle\NavigationBundle\Build\BuildDispatcher` aufgerufen, da werden die Resources gesammelt um via `getResources()` abgerufen zu werden und zwar in der `\Webfactory\Bundle\NavigationBundle\Build\TreeFactory`, um da dann mit dem Baum zusammen in einen `\Symfony\Component\Config\ConfigCacheInterface` gesteckt zu werden. Ich finde diesen Weg arg weit, um schnell zu verstehen, dass hier Resources im Sinne des ConfigCacheInterface gemeint sind. Deshalb hab ich weiterhin das starke Bedürfnis, die `addResource()` als weiter innen liegende Interna in eine die Intention verdeutlichende `addTreeCacheExpiryRule()` (o.ä.) zu wrappen.

mpdude commented on README.md:
```diff
@@ -1,65 +1,169 @@
+
+This project was started at webfactory GmbH, Bonn.
+
+- <http://www.webfactory.de>
```
`https`

mpdude commented on README.md:
```diff
@@ -1,65 +1,169 @@
+- <http://www.webfactory.de>
+- <http://twitter.com/webfactory>
+
+Copyright 2015 webfactory GmbH, Bonn. Code released under [the MIT license](LICENSE).
```
2015 – 2021, wenn Du schon so viel Liebe investierst ;-)

MalteWunsch replied:
Irgendwie finde ich diese Zeile absurd, aber klar, ist angepasst 👍

mpdude commented on Resources/config/services.xml:
```diff
@@ -3,11 +3,6 @@
            xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
-
-    <parameters>
-        <parameter key="webfactory_navigation.default_theme.file">WebfactoryNavigationBundle:Navigation:baseNavigationTheme.html.twig</parameter>
```
Potenzieller BC Break, steht nicht in Deiner sonst sehr gründlichen Liste oben.

mpdude commented on Resources/views/Navigation/navigationBlocks.html.twig:
```diff
@@ -0,0 +1,76 @@
+{% endblock %}
+
+{# visibleNodes \Webfactory\Bundle\NavigationBundle\Tree\Node[] #}
+{# level int #}
```
Ist das die Art, wie die IDE die Hinweise für den folgenden Block versteht?

mpdude commented on Tree/Finder.php:
```diff
@@ -23,8 +23,7 @@ class Finder
     /**
-     * @param Node  $object
-     * @param array $requirements
+     * @param Node $object
```
Können wir da einen richtigen Type Hint verwenden, oder geht das aus Gründen nicht?

mpdude commented on Twig/NavigationExtension.php:
```diff
@@ -2,63 +2,166 @@
+    public static function getSubscribedServices(): array
     {
-        $this->treeFactory = $treeFactory;
+        return [Tree::class];
```
N1ce!

mpdude commented on Twig/NavigationExtension.php:
```diff
@@ -2,63 +2,166 @@
+        int $expandedLevels = 1,
+        string $template = 'WebfactoryNavigationBundle:Navigation:navigation.html.twig'
+    ): string {
+        $root = $this->getRootNode($root);
```
Sollten wir hier nicht lieber sagen, `if (!$root instanceof Node) { $root = $this->findNode($root); }`?

Dann könnte die ganze `getRootNode()` Methode vielleicht entfallen und wir nutzen `findNode()`, die eh schon da ist?

mpdude commented on Twig/NavigationExtension.php:
```diff
@@ -76,19 +179,50 @@ public function getNavigationActiveAtLevel($level)
+    public function getAdditionalNavigationItemClasses(Node $node, array $loop, int $level): string
+    {
+        $baseClasses = [
+            $node->isActiveNode() ? 'a' : 'na',
```
Hm, das packt irgendwelche CSS-Klassennamen in die Twig-Extension, wo sie bestimmt nicht gut auszutauschen sind... :-(. Muss das sein?

Wo lebte das vorher? Was das der Grund, warum das "power set" als Funktion angeboten wurde, damit die Ableitung dieser CSS-Klassen-Kombinationen im Theme passieren konnte?

MalteWunsch replied:
Vorher im Twig Template fürs Theme: https://github.com/webfactory/WebfactoryNavigationBundle/blob/f55fa0cd7f46ee8628f30091b01559060660a6b7/Resources/views/Navigation/baseNavigationTheme.html.twig#L29-L45

Ich _vermute_, dass Team Grün entweder alle diese Klassen fürs Styling möchte - und dann der Einheitlichkeit halber in genau dieser Schreibweise, oder keine - dann kann es den Block

```
{% block navigation_item_class -%}
    ni ni-{{ level }} {{ additional_navigation_item_classes(node, loop, level) }}
{%- endblock -%}
```

entsprechend überschreiben. Oder wollt ihr wie vorher (siehe Link oben) möglich, mit der power_set-Funktion neu benannte Klassen bauen können, @polarbirke @megatronCGN?

polarbirke replied:
Tatsächlich sind wir von den kryptischen Kürzeln seit ~2 Jahren weg und tendieren eher dazu, im projekt-spezifischen Theme verbose Schreibweise inkl. Projekt-Namespace/Präfix zu definieren.

Beispiel: https://github.com/webfactory/jugendfuereuropa/blob/98014_Redesign/src/Webforum/JFE/Bundle/WebsiteBundle/Resources/views/Navigation/main_navigation_theme.html.twig#L68 ff.

mpdude replied:
Die Idee war mal in grauer Vorzeit, Klassen und -kombinationen "für alles" zu haben, so dass man Styling auf ein Element anwenden konnte, wenn es "das aktive Element auf 2. Ebene ist, das nicht das letzte unter seinen Geschwistern darstellt" oder so ähnlich. Es wurde tatsächlich immer nur ein Bruchteil der Potenzmenge (= alle möglichen Kombinationen der Klassen) gebraucht.

Worum es mir geht: Entweder, wir behalten das Feature für BC bei, oder wir machen an der Stelle einen break, lassen die Klassen raus und man muss im Projekt selbst definieren, was man will... das geht sicher auch ohne "power set". 

Verstehe ich @MalteWunsch richtig, dass Du das `power_set` _nicht_ mehr als Twig-Funktion anbieten möchtest, dann aber – um noch die alten Klassen für BC liefern zu können – den Krempel direkt in der Twig-Extension hast?

Wenn dann aber jemand die Klassen überschrieben hat und dafür das `power_set` braucht, dann haben wir nichts gewonnen... die BC ist dann trotzdem gebrochen, und die CSS-Klassen im Extension-Code sind auch wertlos. 🤷🏻‍♂️

MalteWunsch replied:
Potzblitz: [wir haben tatsächlich Klassen mit Hilfe von `power_set` überschrieben](https://github.com/search?q=org%3Awebfactory+power_set&type=code), und das nicht nur einmal. Das hatte ich gar nicht in Betracht gezogen. Alles klar, dann bleibt die `power_set` als TwigFunction erhalten!

Dann schwenke ich auch auf die alte Implementierung des `{% block navigation_item_class -%}`-Blocks zurück. Ich fand die zwar schwerer lesbar, bin da aber sicherlich gebiased. Außerdem gefiel mir auch nicht, dafür eine TwigFunction `additional_navigation_item_classes` public anbieten zu müssen.

mpdude commented on composer.json:
```diff
@@ -4,20 +4,26 @@
             "Webfactory\\Bundle\\NavigationBundle\\": ""
-        }
+        },
+        "exclude-from-classmap": [
```
Das scheint mir hier falsch.

MalteWunsch replied:
Möglich! Ich kam drauf, weil der `composer-require-checker` Klassen aus `vendor` durchsucht und bemängelt hat. Ich vermute, wir haben das Problem sonst nicht, weil wir sonst eher so etwas haben wie

```
"psr-4": {
            "Webfactory\\Bundle\\NavigationBundle\\": "src"
        }
},
```

Vorschlag: alles in `src/` verschieben und Autoloading anpassen?

MalteWunsch replied:
Vorschlag selbst für gut befunden und umgesetzt ^^

mpdude commented on CHANGELOG.md:
```diff
@@ -58,5 +133,51 @@ Changelog
    Damit können Klienten den Baum verändern/erzeugen/ergänzen - und zwar genau in dem Moment, wo er erstmals benötigt wird.
    Details zur Verwendung in 818daa580f243344b183e67438dea3592eb24b8a.
 
+## Version 2
```
Viel besser, das hier im Changelog zu haben! (Vermute, es war vorher in der README?)

MalteWunsch replied:
Genau!

mpdude reviewed (commented):
Vielleicht ist meine Frage allegemeiner, ob/wie wir Parameter-Flexibilität in die Twig-Hilfsfunktionen reinbekommen, oder ob sie eventuell schon von Haus aus da ist?

mpdude commented:
Bin mir gerade nicht sicher, ob Named Parameters für Twig-Funktionen nicht von Haus aus da sind...

MalteWunsch committed:
Bump copyright notice year

MalteWunsch committed:
https URLs

MalteWunsch committed:
Liste entfallenen Parameter

MalteWunsch committed:
Verschiebe Code in src/ (Case 123872)

Damit lässt sich PSR4-Autoloading leichter/passgenauer einrichten, um Verzeichnisse wie vendor und test nicht ins Autoloading mit einzuschließen.

MalteWunsch committed:
Entferne private method getRootNode (Case 123872)

So ist es deutlich kürzer. Die InvalidArgumentException brauchen wir nicht mehr, weil die find-Methode bereits ein ge-type-hint-etes array entgegen nimmt.

MalteWunsch committed:
Empfehle Nutzung von named parameters in Twig (Case 123872)

Das erhöht die Lesbarkeit und wahrt mehr Kontinuität beim Update aufs webfactory/navigation-bundle 4.

MalteWunsch committed:
Stelle TwigFunction power_set wieder her, entferne additional_navigation_item_classes (Case 123872)

MalteWunsch committed:
Bennene addTreeCacheExpiryRule zu addResourcesToTreeCache um (Case 123872)

Es sind keine Regeln, sondern `Symfony\Component\Config\Resource\ResourceInterface`s für das `\Symfony\Component\Config\ConfigCacheInterface`.
